tests/integration/high_availability/test_replication.py

```diff
@@ -0,0 +1,318 @@
+    await clear_writes_action.wait()
+
+
+@pytest.mark.order(1)
+@pytest.mark.replication_tests
+async def test_build_and_deploy(ops_test: OpsTest) -> None:
+    """Simple test to ensure that the mysql and application charms get deployed."""
+    await high_availability_test_setup(ops_test)
+
+
+@pytest.mark.order(2)
+@pytest.mark.abort_on_fail
+@pytest.mark.replication_tests
+async def test_kill_primary_check_reelection(ops_test: OpsTest, continuous_writes) -> None:
+    """Test to kill the primary under load and ensure re-election of primary."""
+    mysql_application_name, _ = await high_availability_test_setup(ops_test)
+
+    mysql_unit = ops_test.model.applications[mysql_application_name].units[0]
+    primary = await get_primary_unit(ops_test, mysql_unit, mysql_application_name)
+    primary_name = primary.name
+
+    # kill the primary pod
+    client = lightkube.Client()
+    client.delete(Pod, primary.name.replace("/", "-"), namespace=ops_test.model.info.name)
+
+    async with ops_test.fast_forward():
+        # wait for model to stabilize, k8s will re-create the killed pod
+        await ops_test.model.wait_for_idle(
+            apps=[mysql_application_name],
+            status="active",
+            raise_on_blocked=True,
+            timeout=TIMEOUT,
+        )
+
+        # ensure a new primary was elected
+        mysql_unit = ops_test.model.applications[mysql_application_name].units[0]
+        new_primary = await get_primary_unit(ops_test, mysql_unit, mysql_application_name)
+        new_primary_name = new_primary.name
+
+        assert primary_name != new_primary_name
+
+        # wait (and retry) until the killed pod is back online in the mysql cluster
+        try:
+            for attempt in Retrying(stop=stop_after_delay(5 * 60), wait=wait_fixed(10)):
+                with attempt:
+                    cluster_status = await get_cluster_status(ops_test, mysql_unit)
+                    online_members = [
+                        label
+                        for label, member in cluster_status["defaultreplicaset"][
+                            "topology"
+                        ].items()
+                        if member["status"] == "online"
+                    ]
+                    assert len(online_members) == 3
+                    break
+        except RetryError:
+            assert False, "Old primary has not come back online after being killed"
+
+    last_written_value = await get_max_written_value_in_database(ops_test, primary)
+
+    for attempt in Retrying(stop=stop_after_delay(2 * 60), wait=wait_fixed(3)):
+        with attempt:
+            # ensure that all units are up to date (including the previous primary)
+            for unit in ops_test.model.applications[mysql_application_name].units:
+                written_value = await get_max_written_value_in_database(ops_test, unit)
+                assert written_value > last_written_value, "Continuous writes not incrementing"
+
```

Comment:
Question on this. (Sorry if this is super obvious)

Say we have the following replicas:
```
A +
B
C
```
`+` - primary

and we write values `1, 2, 3` to this replica set. And then we kill the DB process on `A` and now have
```
A x
B + 
C
```
`+` - primary
`x` - db process gone and not reachable

and we write `4` to this replica set (B&C). After some time `A` comes back up and we have 
```
A 
B + 
C
```
`+` - primary

`last_written_value` should be `4` in this case. Writes are continuing and we write `5` to this replica set (A, B, & C).

```
    for attempt in Retrying(stop=stop_after_delay(2 * 60), wait=wait_fixed(3)):
        with attempt:
            # ensure that all units are up to date (including the previous primary)
            for unit in ops_test.model.applications[mysql_application_name].units:
                written_value = await get_max_written_value_in_database(ops_test, unit)
                assert written_value > last_written_value, "Continuous writes not incrementing"
```
this check should assert True on the former primary (`A`) since `5` is greater than `4`. **But my question is** does it guarantee that `4` is on that replica? 

Again sorry if this is obvious

Reply:
You're right that it does not guarantee that `4` is on the replica. However, the goal of this block of code is to ensure that replication is working from the primary to the replicas and that the continuous writes is active. Doing so will confirm that the old primary that was killed is back online and an active member of the cluster. I am quite certain that once replication is active from the primary to all the replicas, any missing data is going to be transferred over (there are other tests for IST and SST I believe - the spirit of this test as I understand it is to ensure that a new primary is auto elected and that the old primary becomes an active replica). Please let me know if I am mistaken and need to refactor the test

Reply:
Approving after our conversation :)